torchmultimodal/modules/layers/parallel_attention.py

```diff
@@ -0,0 +1,269 @@
+        mlp_expansion_ratio: float = 2.6875,  # 8/3 is param matching
+        qk_normalization: bool = True,
+        projection_dropout: float = 0.0,
+        attention_dropout: float = 0.0,
+        use_group_query_attention: bool = True,
+        num_heads_group_query_attention: int = 1,
+        use_in_projection_bias: bool = True,
+        use_out_projection_bias: bool = True,
+        use_weight_init: bool = True,
+        num_layers: int = 0,
+        use_rms_norm: bool = True,
+        use_rotary_embeddings: bool = False,
+        max_expected_seq_len: int = 2048,  # needed only if using rotary
+    ) -> None:
+        super().__init__()
+
+        version_check = not version.parse(torch.__version__) < version.parse("2.0.0")
+        assert (
+            version_check
+        ), f"Parallel Attention Blocks requires PT 2.0+, you are running {torch.__version__}.\nPlease upgrade your PyTorch version."
+
+        self.num_heads = num_heads
+        self.emb_dim = emb_dimension
+        self.head_dim = head_dimension if head_dimension else emb_dimension // num_heads
+        assert (
+            self.emb_dim % self.num_heads == 0
+        ), f"dimensions {self.emb_dim} must be evenly divisible by num_heads {num_heads=}"
+
+        # group query attn
+        if use_group_query_attention:
+            assert (
+                self.num_heads % num_heads_group_query_attention == 0
+            ), f"{self.num_heads=} not evenly divisible by {num_heads_group_query_attention=}"
+
+        self.use_variable_kv = use_group_query_attention
+        self.group_num_kv = num_heads_group_query_attention
+        self.num_kv = self.group_num_kv if self.use_variable_kv else self.num_heads
+        self.kv_head_dims = self.head_dim * self.num_kv
+        self.kv_expansion_factor = int(self.num_heads / self.group_num_kv)
+        assert (
+            self.kv_expansion_factor > 0
+        ), f"kv expansion factor must be positive integer, got {self.kv_expansion_factor=}"
+
+        self.mlp_hidden_dim = int(mlp_expansion_ratio * self.emb_dim)
+
+        self.qk_norm: bool = qk_normalization
+
+        self.attention_dropout = nn.Dropout(attention_dropout)
+        self.mlp_dropout = nn.Dropout(projection_dropout)
+
+        # weight init
+        self.num_layers = num_layers
+        self.use_weight_init = use_weight_init
+        if self.use_weight_init:
+            assert (
+                self.num_layers > 0
+            ), f"Need to pass in global num layers for weight init, {self.num_layers=}"
+
+        self.weight_init_standard_dev = 0.02 / math.sqrt(2 * self.num_layers)
+
+        self.use_in_projection_bias = use_in_projection_bias
+        self.use_out_projection_bias = use_out_projection_bias
+
+        # previous init params, moved to internal defaults for streamlining
+        normalization_layer = RMSNorm if use_rms_norm else nn.LayerNorm
+        self.mlp_activation = nn.SiLU()
+
+        self.num_q = 1
+
+        self.in_proj_dims = [
```

Comment:
nit: add a comment either here or on self.in_proj definition about the fusing you're doing. I think fusing of the MLP and gate could be a bit unusual for those unfamiliar with parallel attention